File: node/packages/fluent-cards/tests/inputs.test.ts
import { describe, it } from 'node:test';
import assert from 'node:assert/strict';
import {
  AdaptiveCardBuilder,
  InputTextBuilder,
  InputNumberBuilder,
  InputDateBuilder,
  InputTimeBuilder,
  InputToggleBuilder,
  InputChoiceSetBuilder,
  toJson,
  TextInputStyle,
  ChoiceInputStyle,
} from 'fluent-cards';
import type { InputText, InputNumber, InputDate, InputTime, InputToggle, InputChoiceSet } from 'fluent-cards';

describe('InputTextBuilder', () => {
  it('builds an InputText with all properties', () => {
    const input = new InputTextBuilder()
      .withId('name')
      .withLabel('Your Name')
      .withPlaceholder('Enter name...')
      .withValue('Alice')
      .withMaxLength(100)
      .withIsMultiline()
      .withStyle(TextInputStyle.Email)
      .withRegex('[a-z]+')
      .withIsRequired()
      .withErrorMessage('Required')
      .build();

    assert.equal(input.type, 'Input.Text');
    assert.equal(input.id, 'name');
    assert.equal(input.label, 'Your Name');
    assert.equal(input.placeholder, 'Enter name...');
    assert.equal(input.value, 'Alice');
    assert.equal(input.maxLength, 100);
    assert.equal(input.isMultiline, true);
    assert.equal(input.style, TextInputStyle.Email);
    assert.equal(input.regex, '[a-z]+');
    assert.equal(input.isRequired, true);
    assert.equal(input.errorMessage, 'Required');
  });

  it('type discriminant is Input.Text', () => {
    const input = new InputTextBuilder().withId('x').build();
    assert.equal(input.type, 'Input.Text');
  });
});

describe('InputNumberBuilder', () => {
  it('builds an InputNumber with constraints', () => {
    const input = new InputNumberBuilder()
      .withId('age')
      .withLabel('Age')
      .withMin(0)
      .withMax(120)
      .withValue(25)
      .build();

    assert.equal(input.type, 'Input.Number');
    assert.equal(input.id, 'age');
    assert.equal(input.min, 0);
    assert.equal(input.max, 120);
    assert.equal(input.value, 25);
  });
});

describe('InputDateBuilder', () => {
  it('builds an InputDate', () => {
    const input = new InputDateBuilder()
      .withId('dob')
      .withMin('2000-01-01')
      .withMax('2024-12-31')
      .withValue('2024-01-01')
      .build();

    assert.equal(input.type, 'Input.Date');
    assert.equal(input.min, '2000-01-01');
    assert.equal(input.max, '2024-12-31');
    assert.equal(input.value, '2024-01-01');
  });
});

describe('InputTimeBuilder', () => {
  it('builds an InputTime', () => {
    const input = new InputTimeBuilder()
      .withId('startTime')
      .withMin('09:00')
      .withMax('17:00')
      .withValue('10:00')
      .build();

    assert.equal(input.type, 'Input.Time');
    assert.equal(input.min, '09:00');
    assert.equal(input.max, '17:00');
    assert.equal(input.value, '10:00');
  });
});

describe('InputToggleBuilder', () => {
  it('builds an InputToggle', () => {
    const input = new InputToggleBuilder()
      .withId('accept')
      .withTitle('Accept Terms')
      .withValueOn('yes')
      .withValueOff('no')
      .withWrap()
      .build();

    assert.equal(input.type, 'Input.Toggle');
    assert.equal(input.id, 'accept');
    assert.equal(input.title, 'Accept Terms');
    assert.equal(input.valueOn, 'yes');
    assert.equal(input.valueOff, 'no');
    assert.equal(input.wrap, true);
  });
});

describe('InputChoiceSetBuilder', () => {
  it('builds an InputChoiceSet with choices', () => {
    const input = new InputChoiceSetBuilder()
      .withId('hobbies')
      .isMultiSelect()
      .withStyle(ChoiceInputStyle.Expanded)
      .addChoice('Reading', 'reading')
      .addChoice('Gaming', 'gaming')
      .addChoice('Sports', 'sports')
      .build();

    assert.equal(input.type, 'Input.ChoiceSet');
    assert.equal(input.id, 'hobbies');
    assert.equal(input.isMultiSelect, true);
    assert.equal(input.style, ChoiceInputStyle.Expanded);
    assert.equal(input.choices!.length, 3);
    assert.equal(input.choices![0].title, 'Reading');
    assert.equal(input.choices![0].value, 'reading');
  });

  it('sets choices.data with a Data.Query payload', () => {
    const input = new InputChoiceSetBuilder()
      .withId('people-picker')
      .withChoicesData('graph.microsoft.com/users')
      .build();

    const choicesData = input['choices.data'];
    assert.ok(choicesData, 'choices.data should be present');
    assert.equal(choicesData!.type, 'Data.Query');
    assert.equal(choicesData!.dataset, 'graph.microsoft.com/users');
  });
});

describe('AdaptiveCardBuilder input methods', () => {
  it('adds all input types to body in order', () => {
    const card = AdaptiveCardBuilder.create()
      .addInputText((i) => i.withId('name').withLabel('Name'))
      .addInputNumber((i) => i.withId('age').withLabel('Age'))
      .addInputDate((i) => i.withId('date').withLabel('Date'))
      .addInputTime((i) => i.withId('time').withLabel('Time'))
      .addInputToggle((i) => i.withId('toggle').withTitle('Accept'))
      .addInputChoiceSet((i) => i.withId('choice').addChoice('A', 'a'))
      .build();

    assert.equal(card.body!.length, 6);
    assert.equal(card.body![0].type, 'Input.Text');
    assert.equal(card.body![1].type, 'Input.Number');
    assert.equal(card.body![2].type, 'Input.Date');
    assert.equal(card.body![3].type, 'Input.Time');
    assert.equal(card.body![4].type, 'Input.Toggle');
    assert.equal(card.body![5].type, 'Input.ChoiceSet');
  });

  it('serializes input type discriminants correctly', () => {
    const card = AdaptiveCardBuilder.create()
      .addInputText((i) => i.withId('t1'))
      .addInputChoiceSet((i) => i.withId('cs1').addChoice('A', 'a'))
      .build();

    const json = toJson(card);
    assert.ok(json.includes('"type": "Input.Text"'));
    assert.ok(json.includes('"type": "Input.ChoiceSet"'));
  });
});

File: go/fluentcards/inputs_test.go
package fluentcards_test

import (
	"testing"

	"github.com/rido-min/FluentCards/go/fluentcards"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func TestInputTextBuilder(t *testing.T) {
	t.Parallel()
	card := fluentcards.NewAdaptiveCardBuilder().
		AddInputText(func(it *fluentcards.InputTextBuilder) {
			it.WithID("name").
				WithLabel("Your Name").
				WithPlaceholder("Enter name").
				WithMaxLength(100).
				WithIsMultiline(false).
				WithStyle(fluentcards.TextInputStyleEmail).
				WithIsRequired(true).
				WithErrorMessage("Name is required")
		}).Build()
	el := card["body"].([]any)[0].(map[string]any)
	assert.Equal(t, "Input.Text", el["type"])
	assert.Equal(t, "name", el["id"])
	assert.Equal(t, "Your Name", el["label"])
	assert.Equal(t, "Enter name", el["placeholder"])
	assert.Equal(t, 100, el["maxLength"])
	assert.Equal(t, false, el["isMultiline"])
	assert.Equal(t, "email", el["style"])
	assert.Equal(t, true, el["isRequired"])
	assert.Equal(t, "Name is required", el["errorMessage"])
}

func TestInputNumberBuilder(t *testing.T) {
	t.Parallel()
	card := fluentcards.NewAdaptiveCardBuilder().
		AddInputNumber(func(in *fluentcards.InputNumberBuilder) {
			in.WithID("qty").
				WithLabel("Quantity").
				WithMin(1).
				WithMax(100).
				WithValue(10)
		}).Build()
	el := card["body"].([]any)[0].(map[string]any)
	assert.Equal(t, "Input.Number", el["type"])
	assert.Equal(t, "qty", el["id"])
	assert.Equal(t, float64(1), el["min"])
	assert.Equal(t, float64(100), el["max"])
	assert.Equal(t, float64(10), el["value"])
}

func TestInputDateBuilder(t *testing.T) {
	t.Parallel()
	card := fluentcards.NewAdaptiveCardBuilder().
		AddInputDate(func(id *fluentcards.InputDateBuilder) {
			id.WithID("start").
				WithLabel("Start Date").
				WithMin("2025-01-01").
				WithMax("2026-12-31").
				WithValue("2025-06-15")
		}).Build()
	el := card["body"].([]any)[0].(map[string]any)
	assert.Equal(t, "Input.Date", el["type"])
	assert.Equal(t, "start", el["id"])
	assert.Equal(t, "2025-01-01", el["min"])
	assert.Equal(t, "2026-12-31", el["max"])
	assert.Equal(t, "2025-06-15", el["value"])
}

func TestInputTimeBuilder(t *testing.T) {
	t.Parallel()
	card := fluentcards.NewAdaptiveCardBuilder().
		AddInputTime(func(it *fluentcards.InputTimeBuilder) {
			it.WithID("meeting-time").
				WithMin("09:00").
				WithMax("17:00")
		}).Build()
	el := card["body"].([]any)[0].(map[string]any)
	assert.Equal(t, "Input.Time", el["type"])
	assert.Equal(t, "meeting-time", el["id"])
	assert.Equal(t, "09:00", el["min"])
	assert.Equal(t, "17:00", el["max"])
}

func TestInputToggleBuilder(t *testing.T) {
	t.Parallel()
	card := fluentcards.NewAdaptiveCardBuilder().
		AddInputToggle(func(it *fluentcards.InputToggleBuilder) {
			it.WithID("agree").
				WithTitle("I agree to the terms").
				WithValueOn("true").
				WithValueOff("false").
				WithWrap(true)
		}).Build()
	el := card["body"].([]any)[0].(map[string]any)
	assert.Equal(t, "Input.Toggle", el["type"])
	assert.Equal(t, "agree", el["id"])
	assert.Equal(t, "I agree to the terms", el["title"])
	assert.Equal(t, "true", el["valueOn"])
	assert.Equal(t, "false", el["valueOff"])
	assert.Equal(t, true, el["wrap"])
}

func TestInputChoiceSetBuilder(t *testing.T) {
	t.Parallel()
	card := fluentcards.NewAdaptiveCardBuilder().
		AddInputChoiceSet(func(ics *fluentcards.InputChoiceSetBuilder) {
			ics.WithID("color").
				WithLabel("Favorite Color").
				WithStyle(fluentcards.ChoiceInputStyleExpanded).
				AddChoice("Red", "red").
				AddChoice("Green", "green").
				AddChoice("Blue", "blue")
		}).Build()
	el := card["body"].([]any)[0].(map[string]any)
	assert.Equal(t, "Input.ChoiceSet", el["type"])
	assert.Equal(t, "color", el["id"])
	assert.Equal(t, "expanded", el["style"])
	choices := el["choices"].([]any)
	require.Len(t, choices, 3)
	assert.Equal(t, "Red", choices[0].(map[string]any)["title"])
	assert.Equal(t, "red", choices[0].(map[string]any)["value"])
}
